### DISCUSSION TITLE
Delimited Continuations

### DISCUSSION THREAD

**Discussion by rssh** [rssh/dotty-cps-async#47] [Ideas]
hi sorry for being stranger, went into surgery, now catching up with different things, I just read about multicore ml, how they use fibers to implement their single shot delimited continuation.

cats-effect, zio, monix and even kotlin coroutine and later loom have something close to fibers, do you think we can use those as a runtime for continuation, thereby have a delimited continuation system in scala, I think you already did most of the hard work.

once we have single shot delimited continuation, we can build the generators, resumable exception, effect handlers, and so forth.

I am just shooting some ideas, I could be wrong.

_Originally posted by @renghen in https://github.com/rssh/dotty-cps-async/discussions/39#discussioncomment-1045678_

**Comment by rssh** [comment]
It's quite an interesting theme, from the other side -- I can't find a good practical use-case, because for all cases, which I trying to invent, exists analog based on monadic context.  On the other side -- I'm not sure, that I fully understand continuations.  I.e. is this some monad, which allows in runtime to apply effects, then how does this differ from the combination of Monad + Effect handlers embedded in some framework (like ZLayer) [?].   If not - what is the essential difference (?).  Run handlers within one monad (?) i.e. decouple the definition of handlers from monad?  
Maybe this is the ability to call function N-tymes (multi-shot).  
From reading effekt paper: https://github.com/b-studios/scala-effekt
[Effect: Capability-passing style for type- and effect-safe, extensible effect handlers in Scala](https://www.cambridge.org/core/journals/journal-of-functional-programming/article/effekt-capabilitypassing-style-for-type-and-effectsafe-extensible-effect-handlers-in-scala/A19680B18FB74AD95F8D83BC4B097D4F)  I still can't understand a perfect use-case. [except typing capacities].

Maybe slow but deterministic way will try to implement from scratch continuation monad, based on
https://www.microsoft.com/en-us/research/wp-content/uploads/2005/01/jfp-revised.pdf
And in the process discover - what exactly will be the use-case and what is missing in the current monadic interfaces.

**Reply by rssh** [reply]
After some thinking, I can try to answer themself
 - The sense is to have one Control monad, which will be 'the main monad',  and therefore all Control[A]. will be compatible with each other, and we can stack them.
 -  But how do we interpret the evaluation of Control[A]?    If this is something like Rand[A]. or Search[A]?   Effekt (as I understood)  choose to have canonical evaluation for Control[A, StandardCapacities]. and effect handlers are something like mappings: Control[A, Effects | MyEffect] => Control[ MyEffect.Result[A] , Effect].     
 - In real life, we also have a platform and execution [compile-time,  prepare-time [for something like IO],  runtime-unsafe]

**Reply by renghen** [reply]
I think should concentrate on the delimited continuation part, and from there on we can build other constructs. In multicore ocaml
from what I read(https://arxiv.org/pdf/2104.00250.pdf) in section 4.2.1, A **continuation** 𝑘 is either empty [] or a non-empty list of
fibers. A **fiber** 𝜑 B (𝜓, 𝜂) is a list of frames 𝜓 and a handler
closure 𝜂 B (ℎ, 𝜖), which is a pair of handler ℎ and its environment 𝜖.

Once we have a dynamic runtime for continuation, with an interface say CONT, we can start to use macros to build stuff that can be written in continuation such as generators, exception, effect handlers etc..

**Reply by rssh** [reply]
Agree.

**Reply by rssh** [reply]
From the other side - generators can be built on top of async/await easier,  exception is build-in.  So, a question - what is an example of application of delimited continuation [except this is interesting as-is]. is still open.

**Reply by renghen** [reply]
a better exception mechanism martin is is looking closely to effect handlers for that, from there we can provide resumable exception, effect handlers.

With continuation, we can have monads/monad stack written in a more direct style. Monad reflection for example solves a lot of the ergonomics issues of monad stack, but needs a continuation mechanism.

I think if we can get a proper continuation mechanism with a plug able runtime, this can be applied to scala.js, scala native, who would be without continuation mechanism even if loom comes out, as loom is a jvm centric thing.

**Reply by raulraja** [reply]
Hi 👋 , relevant to this discussion, this is more or less a walkthrough on how continuations are implemented in Kotlin https://jorgecastillo.dev/digging-into-kotlin-continuations . IMO There isn't really a need for a fiber runtime if that implies any sort of async or context scheduling. Continuations at their base can all be desugared into loops and labeled states. Kotlin builds the async and stream runtime as separate libs on top of these internals shown in the post. The compiler desugars `suspend fun` as a function with an additional continuation argument that is threaded through the suspension loop state machine. Most of the time you get a single allocation for the Continuation callback but sync cases under suspension can be treated specially with the same overhead as regular functions.

**Reply by renghen** [reply]
agreed, continuation can be implemented with or without a runtime support.

Both approaches will require compiler magic thru macros or otherwise.

The question is which route is best, the delimited continuation of Scala is now async plugin for scala 2,

**Reply by rssh** [reply]
so, one real application is still 'effect handlers' which is needed for  what(?)
  //  [resumable exception can be a cool example, but VB practice shows that this has not become a  practically useful idiom.]
Yet one possible example:  resource closing at the end of the scope.  

Let's make table:


| Feature          |for end-user | for library-builder | on async |   on delimited continuation | link | interest vect| comment|
|--------------|-------------|-------------------|----------|----------------------------|----|-------------|----------| 
| generators     |                1    |  1                             | 1              |  1                                            |  #47  | 1   |.    |
|. resource closing |        1.   |.   1                            |  1.            |.   1                                          |         |.      |.     |
| effect handler |             0     |  1                             |  0            |  1                                            |          |   ?  |. recursion for eu.|
|  resumable exceptions |. 1  |.  ?                            |. 0            |   1                                           |          |   | no more then joke. |
|  monadic reflection.     | 0  |. 1.                             |  ?             |  1                                           |   http://cs.ioc.ee/mpc-amast06/msfp/filinski-slides.pdf |   |  technique for effect handlers[?]  |
| probabilistic programming | 1 | 0 |  1  |. 1. |.   |   ?   |    |
| non-local return| 1 | 0 |  0 | 1 |. |.  | unclear how interface this in macro | 
|.  Question? | 1 |   * |  0  |  1. |.   |. |. Question ? |  
----

Is exists something which can be put instead 'Question'  on this table ?   I.e. something which is not implementable on top of plain monadic async and useful for programmers.

**Reply by rssh** [reply]
Another interesting journey will be trying  to
-  build CpsTryMonad for. `cats.data.ContT`.  (which is a port of a Haskell Continuation Monad [not delimited, async is one outer scope. ])
```
  [X] =>> cats.data.ContT[IO,X]
```
- inject `IO`  and `[X] =>> Resource[IO,X]`. here

this as can be a minimal practical implementation of the close idea in https://github.com/rssh/cps-async-connect

**Reply by renghen** [reply]
the ContT is not recommended, and cats effects is more in favor, hence my original idea of using fibers as a basics for cps.

One use of cps is also to provide direct style development as it is shown in your work for async await, with cps I think it can extended to other areas, as you have mention a few above.

my idea of using cps, is to make monad reflection a thing, effect handlers also, so as to make the programming with effects easier in scala and more ergonomics, instead of going to zio like super monads.

All the building blocks are there, so why not combine them, in a more ergonomic way using cps.

**Reply by renghen** [reply]
from my talk with Daniel on the cats effect discord channel,

IO.async (or IO.cont) is shift in scheme.

So we only need syntactic sugar on top of that.

now is my question is the async await enough ? 

or do we need something more?

**Reply by rssh** [reply]
I still can't understand, what your want to build.  Having a concrete example (maybe in form of pseudocode, etc, can help)
 
[Even if you have 'direction' instead 'goal',  setting a temporary concrete goal and achieving it will give you information, necessary for understanding the area, setting the next concrete goal, and formulate a global goal].

If you want to use IO.cont, then prepare an example which should be better to write in a direct style instead cont.  [But as I remember, cont is an internal function, which lives in the 'Async' interpreter].  

You want to have canonical  shift / reset. -- write a code example with one inside async
- what typing should be for shift()?
-  are we need reset() or async will give us bounds() ?   If yes - what typing will be in reset()?
- what is the relationship between the types of async() and reset()?

Starting from this we will be able to understand the concrete architecture of such construction and think about how to build one.

**Reply by renghen** [reply]
Sorry for the confusion.

My point is as follows, we have many IO effect system in scala, from those we can build continuations construct/syntax in scala.

your work have already proven we can do async/await in a very simple way to any IO effect monad.

What I want to build is a full single shot multi prompt delimited continuation using tools such as cats effect fibers, cats fibers from daniel 's explanation are continuations, and that IO.async(IO.cont) is the equivalent of a shift(which I only learned yesterday)

My question is simple then, can we using those information build a reset/shift or a continue k a la multicore ml (https://github.com/ocamllabs/ocaml-effects-tutorial)

PS Raul does point out that we do **NOT** need a fiber system, also fibers are already continuation in cats.

**Reply by raulraja** [reply]
@renghen I actually meant there is no need to include any kind of fiber system. A compiler plugin can desugar functions that refer to continuations without fibers by transforming async/suspended/continuation functions to an efficient state machine. something like this https://gist.github.com/raulraja/3f339c00ed4b5e25330f4e80700bf91b for two simple functions, one that is sync and another that has an async completable callback.  That is decompiled from a Kotlin program with suspended functions.

**Reply by renghen** [reply]
yes, you are right I mean to say we do **not** need a fiber system, but typo error sorry for that. I corrected it.

**Comment by rssh** [comment]
Btw,   classical  examples of the delimited continuations monad are added to tests:
 https://github.com/rssh/dotty-cps-async/blob/master/shared/src/test/scala/cpstest/CpsContT.scala
